source/Patches/AmongUsClient_OnGameEnd.cs

```diff
@@ -10,8 +10,29 @@ namespace TownOfUs
     [HarmonyPatch(typeof(EndGameManager), nameof(EndGameManager.Start))]
     public class EndGameManager_SetEverythingUp
     {
+        // Implement a method to record the roles assigned to each player at the end of a game.
+        // Store the role history in a session-based collection.
+        private static Dictionary<byte, List<RoleEnum>> _sessionRoleHistory = new Dictionary<byte, List<RoleEnum>>();
+
+        public static void RecordSessionRoles() {
+            foreach (var player in PlayerControl.AllPlayerControls) {
+                var playerId = player.PlayerId;
+                var roleType = Role.GetRole(player)?.RoleType ?? RoleEnum.None;
+
+                if (!_sessionRoleHistory.ContainsKey(playerId)) {
+                    _sessionRoleHistory[playerId] = new List<RoleEnum>();
+                }
+
+                _sessionRoleHistory[playerId].Add(roleType);
+            }
+        }
+
         public static void Prefix()
         {
+          if (CustomGameOptions.WeightedRoleSelection)  
+          { 
+              RecordSessionRoles(); // Record roles at the end of each game
+          }
```

Comment:
This should probably happen at the start of the game and not the end as the final roles people end up with are not the same as those they initially were assigned and we only care about those original roles.

Additionally, I'm not sure how PlayerIds are handled in Among Us so doing it at game start may handle situations like disconnects better?

Reply:
well, no this affect end game because you can see there is a harmony patch to patch the end game 
```suggestion
        [HarmonyPatch(typeof(EndGameManager), nameof(EndGameManager.Start))]
```
and about player id is a bit complicated

i don't think so that   will make disconnects better

Reply:
I understand that currently it's targeting the end of the game, I'm suggesting that instead targeting the start of the game would make more sense due to the switching of roles and potentially disconnects (though the latter point you're saying doesn't matter). Conceptually it still makes more sense to me to save this stuff pretty much the moments it is generated

Reply:
what do you mean by "genereted" ?there

Reply:
When the original roles are randomly rolled and assigned.